### PR TITLE
Refactored annotation values

### PR DESCRIPTION
**Pull request by awanlin**
## Hey, I just made a Pull Request!

Consolidated getting the annotation values into a single function to help with future changes, specifically adding multi-org support

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
Hi @marleypowell, was wondering if you could give this a review? Basically there's now support for multi-orgs and I'm slowly working towards this. This seemed like a small chunk I could break out first to get things rolling.

**Comment by awanlin** [comment]
Sorry, while working through https://github.com/backstage/backstage/pull/19620 I realized I'm going to need the provide the `host` as well as the `org` so marking this as a Draft for now. Intend to fix this up over the weekend.

**Comment by awanlin** [comment]
Alright, ready for review 👍

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+} {
+  const hostOrgAnnotation =
+    entity.metadata.annotations?.[AZURE_DEVOPS_HOST_ORG_ANNOTATION];
+  const { host, org } = getHostOrg(hostOrgAnnotation);
+
+  const projectRepoAnnotation =
+    entity.metadata.annotations?.[AZURE_DEVOPS_REPO_ANNOTATION];
+  if (projectRepoAnnotation) {
+    const { project, repo } = getProjectRepo(projectRepoAnnotation);
+    const definition = undefined;
```
Comment (marleypowell) [inline plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:41]: Is there much need for this declaration?

**Review by marleypowell — CHANGES_REQUESTED** [review]
Nice one, just a few small suggestions

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        kind: 'Component',
+        metadata: {
+          namespace: 'default',
+          name: 'project-repo',
+          annotations: {
+            'dev.azure.com/project-repo': 'projectName/repoName',
+          },
+        },
+      };
+      const { project, repo, definition, host, org } =
```
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.test.ts]: Yeah, had not considered that, I've refactored the tests to use `toEqual` on the return object 👍

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+export function getAnnotationValuesFromEntity(entity: Entity): {
+  project: string;
+  repo?: string;
+  definition?: string;
+  host?: string;
+  org?: string;
+} {
+  const hostOrgAnnotation =
+    entity.metadata.annotations?.[AZURE_DEVOPS_HOST_ORG_ANNOTATION];
```
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts]: Sure, did a little juggling and this is in place now 👍

File: plugins/azure-devops/src/hooks/useGitTags.ts [hunk 4]
```diff
@@ -20,18 +20,21 @@ import { Entity } from '@backstage/catalog-model';
...
-  const { project, repo } = useProjectRepoFromEntity(entity);
+  const { project, repo } = getAnnotationValuesFromEntity(entity);
 
-  const { value, loading, error } = useAsync(() => {
-    return api.getGitTags(project, repo);
+  const { value, loading, error } = useAsync(async () => {
+    if (repo) {
+      return await api.getGitTags(project, repo);
+    }
+    return undefined;
```
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/azure-devops/src/hooks/useGitTags.ts:37]: Yes, I can do that. I'm having a bit of a hard time with this. Any chance you know of some examples of testing hooks?

[the remark this answers is not in the document]
Comment (freben) [reply plugins/azure-devops/src/hooks/useGitTags.ts:37]: https://github.com/backstage/backstage/blob/master/plugins/catalog-react/src/hooks/useEntity.test.tsx#L32

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  org?: string;
+} {
+  const annotation = annotations?.[AZURE_DEVOPS_HOST_ORG_ANNOTATION];
+  if (!annotation) {
+    return { host: undefined, org: undefined };
+  }
+
+  if (!annotation.includes('/')) {
+    throw new Error(
+      `Value for annotation ${AZURE_DEVOPS_HOST_ORG_ANNOTATION} was not in the correct format: <host-name>/<organization-name>`,
```
Comment (freben) [inline plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:106]: Suggestion: This error message could do well to include the actual value found; that makes the source of the error easier to locate. Could also refactor to have positive tests and a single throw

```ts
if (annotation.includes('/')) {
  if (...) {
    return;
  }
}

throw new Error(`Invalid value for annotation '${AZURE_DEVOPS_HOST_ORG_ANNOTATION}'; expected format X, found '${annotation}'`);
```

Same for the others above

**Review by marleypowell — APPROVED** [review]

**Comment by freben** [comment]
@awanlin let me know when ready!

**Comment by awanlin** [comment]
Thanks @freben, busy working on this, slow going with the tests but making progress 👍

**Comment by freben** [comment]
Hey we're still waiting for a bit more stuff to go into this one right?

**Comment by awanlin** [comment]
Yeah, sorry @freben, I have not had a lot of free time to work on this :(

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts (hunk 5 above)
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts]: Good call, cleaned up the code as well, thanks!

File: plugins/azure-devops/src/hooks/useGitTags.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/azure-devops/src/hooks/useGitTags.ts]: I've added tests for all the hook I changed to use `getAnnotationValuesFromEntity`. I also took out this if check as if this is missing it would already have thrown an error before this.

One thing I ran into with the test was that I needed to use a `tsx` extension but my hook are just `ts`, feel like I might be missing something obvious. Advice welcome

File: plugins/azure-devops/src/components/ReadmeCard/ReadmeCard.tsx [hunk 6]
```diff
@@ -87,16 +87,14 @@ export const ReadmeCard = (props: Props) => {
...
-        project,
-        repo,
-      }),
-    [api, project, repo, entity],
-  );
+  const { loading, error, value } = useAsync(async () => {
+    if (repo) {
+      return await api.getReadme({ project, repo });
+    }
+    return undefined;
```
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/azure-devops/src/components/ReadmeCard/ReadmeCard.tsx]: Alright, refactored this into a hook like all the others 👍

**Comment by awanlin** [comment]
@freben, finally got some time to pick this up and it's read for a review, please

File: plugins/azure-devops/src/hooks/useReadme.test.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+          'dev.azure.com/project-repo': 'fake',
+        },
+      },
+    };
+
+    expect(() =>
+      renderHook(() => useReadme(entity), {
+        wrapper: Wrapper,
+      }),
+    ).toThrow(
```
Comment (freben) [inline plugins/azure-devops/src/hooks/useReadme.test.tsx:104]: So the hook can fail in two distinct ways: throwing synchronously, or returning an error asynchronously. This goes for the other ones too. That's probably fine if it's intentional (they are in a sense two different classes of error), but I wanted to highlight that it's something that consumers might be surprised by. I know these hooks are for internal use, but that immediate throw is one that's more likely to bubble immediately up to the react root and give a more violent crash than whatever handling is in place for the "calm" return of an Error.

So in summary - might it make sense to move all `getAnnotationValuesFromEntity` calls into the `useAsync` to unify how errors are reported?

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  host?: string;
+  org?: string;
+} {
+  const annotation = annotations?.[AZURE_DEVOPS_HOST_ORG_ANNOTATION];
+  if (!annotation) {
+    return { host: undefined, org: undefined };
+  }
+
+  if (annotation.includes('/')) {
+    const [host, org] = annotation.split('/');
```
Comment (freben) [inline plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:94]: this silently rejects anything past a second slash - no big deal, just mentioning that it's not super strict on input

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  project?: string;
+  repo?: string;
+} {
+  const annotation = annotations?.[AZURE_DEVOPS_REPO_ANNOTATION];
+  if (!annotation) {
+    return { project: undefined, repo: undefined };
+  }
+
+  if (annotation.includes('/')) {
+    const [project, repo] = annotation.split('/');
```
Comment (freben) [inline plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:73]: this silently rejects anything past a second slash - no big deal, just mentioning that it's not super strict on input

**Review by freben — APPROVED** [review]
Alright, happy to approve in current state, just some optional comments on possible (future?) improvements

**Comment by awanlin** [comment]
Thanks @freben 🚀

File: plugins/azure-devops/src/hooks/useReadme.test.tsx (hunk 7 above)
Comment (awanlin) [reply plugins/azure-devops/src/hooks/useReadme.test.tsx:104]: Yes, that was something I was thinking about, will do that in a follow up PR 👍

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts (hunk 8 above)
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:94]: Oh, then I think maybe I need to revisit this and add some more tests. It's possible for some to have a host like `company.com/tsf`. I forgot about that case. Thankfully I'm doing this work in small chunks and these aren't being used yet. Will open a follow up PR for this. Thanks for this comment!

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts (hunk 9 above)
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:73]: This one we are safe, you can not use `/` in project or repo names 👍

Comment (freben) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:73]: oh but you know, accidents etc

Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:73]: Right, will think on this as well

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts (hunk 8 above)
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:94]: https://github.com/backstage/backstage/pull/21797 resolves this

File: plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts (hunk 9 above)
Comment (awanlin) [reply plugins/azure-devops/src/utils/getAnnotationValuesFromEntity.ts:73]: https://github.com/backstage/backstage/pull/21797 resolves this too
